File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/country/CountryResponse.java
package openfoodfacts.github.scrachx.openfood.models.entities.country;

import java.util.ArrayList;
import java.util.Map;

import openfoodfacts.github.scrachx.openfood.network.ApiFields;

/**
 * Created by Lobster on 04.03.18.
 */
// MUST represent the entire JSON object as given by the API
public class CountryResponse {
    private final Map<String, String> names;
    private final Map<String, String> cc2;
    private final Map<String, String> cc3;
    private final String tag;

    public CountryResponse(String code, Map<String, String> names, Map<String, String> cc2, Map<String, String> cc3) {
        this.tag = code;
        this.names = names;
        this.cc2 = cc2;
        this.cc3 = cc3;
    }

    public Country map() {
        Country country = new Country(tag,
            new ArrayList<>(),
            cc2.get(ApiFields.Defaults.DEFAULT_TAXO_PREFIX),
            cc3.get(ApiFields.Defaults.DEFAULT_TAXO_PREFIX));
        for (Map.Entry<String, String> name : names.entrySet()) {
            country.getNames().add(new CountryName(country.getTag(), name.getKey(), name.getValue()));
        }

        return country;
    }

}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/country/CountryResponse.kt
package openfoodfacts.github.scrachx.openfood.models.entities.country

import openfoodfacts.github.scrachx.openfood.models.entities.EntityResponse
import openfoodfacts.github.scrachx.openfood.network.ApiFields

/**
 * Created by Lobster on 04.03.18.
 */
// MUST represent the entire JSON object as given by the API
class CountryResponse(
        private val tag: String,
        private val names: Map<String, String>,
        private val cc2: Map<String, String>,
        private val cc3: Map<String, String>
) : EntityResponse<Country> {
    override fun map() = Country(tag,
            names.map { CountryName(tag, it.key, it.value) },
            cc2[ApiFields.Defaults.DEFAULT_TAXO_PREFIX],
            cc3[ApiFields.Defaults.DEFAULT_TAXO_PREFIX])
}